static_frame/core/bus.py

```diff
@@ -739,135 +739,142 @@ def rehierarch(self,
-                    store=self._store,
-                    config=self._config,
-                    labels=(label for label, f in zip(target_labels, target_values) # type: ignore
-                            if f is FrameDeferred),
-                    max_persist=self._max_persist,
-                    )
-            targets_items = zip(target_labels, target_values) # type: ignore
+        # NOTE: prepare iterable of pairs of label, Frame / FrameDeferred; ensure that for every FrameDeferred, the appropriate Frame is loaded and yielded from the store_reader in order. We must ensure within the target of requested Frame we do not delete any previously-loaded Frame. If max_persist is less than the target, reduce the target to max_persist.
 
-        # Iterate over items that have been selected; there must be at least 1 FrameDeffered among this selection
+        if key_is_element:
+            store_reader = iter((self._store.read(target_labels, config=self._config[target_labels]),)) # type: ignore
+            targets_items = ((target_labels, target_values),) # type: ignore
+        # more than one Frame
+        elif (not max_persist_active
+                or max_persist == 1
+                or loaded_needed <= loaded_available # pyright: ignore
+                ):
+            # only read-in labels that are deferred; as loaded_needed is less than loaded_available, no Frame will be removed
+            if target_loaded_count:
+                labels_to_read = target_labels[~target_loaded]
+            else: # no targets are loaded
+                labels_to_read = target_labels
+
+            store_reader = self._store.read_many(labels_to_read, config=self._config)
+            targets_items = zip(target_labels, target_values)
+        # max_persist_active, must delete some Frame
+        else:
+            if loaded_needed <= max_persist:
+                # loaded_needed is less than _max_persist but greater than loaded_available, meaning that some Frame have to be deleted. we must ensure we do not delete a Frame we already have loaded within the target region, so move them to the back of the LRU
+                if target_loaded_count:
+                    # update LRU position to ensure we do not delete in target
+                    for label in target_labels[target_loaded]: # update LRU position
+                        self._last_accessed[label] = self._last_accessed.pop(label, None)
+                    labels_to_read = target_labels[~target_loaded]
+                else: # no targets are loaded
+                    labels_to_read = target_labels
+
+            else: # loaded_needed > max_persist:
+                # Need to load more than max_persist, so limit to last max_persist-length components. All other Frame, if loaded, will be deleted
+                # assert max_persist < len(target_labels)
+                target_labels = target_labels[-max_persist:] # type: ignore # pylint: disable=E1130
```

Comment:
I think this is the optimization, right? If max_persist is shorter than the number of items needed, there's no reason to load the initial frames, since they will immediately be dropped

Reply:
Yes, and an important one. Previously, if the selection range was greater than `max_persist`, we would have uselessly loaded and then dropped (potentially many) `Frame`!